### PR TITLE
[stdlib] Add `unreachable` check instead of `constrained[False]`

### PR DESCRIPTION
**Pull request by Benny-Nottonson**
As mentioned in #2547, there should realistically be an `unreachable` function, to indicate when a segment of code can never be reached. 

For example
```mojo
fn one_or_two[value: Int]() -> Int:
    if value == 1:
        return 1
    elif value == 2:
        return 2
    unreachable()
```

The above code demonstrates a great use case, where the alternative now would be:

```mojo
fn one_or_two[value: Int]() -> Int:
    constrained[value == 1 or value == 2]()
    if value == 1:
        return 1
    return 2
```

This is problematic in my opinion as it creates two issues:
1. You are now checking `value == 1` two separate times, the only way to avoid this is creating an unnecessary variable.
2. You are reducing readability and harming code structure by leading with the entire body of a conditional.

More information and further suggestions found at https://github.com/modularml/mojo/issues/2547 

Signed-off-by: benny-nottonson <bennynottonson@gmail.com>

### CODE REVIEW

File: stdlib/src/builtin/constrained.mojo [hunk 1]
```diff
@@ -33,3 +33,16 @@ fn constrained[cond: Bool, msg: StringLiteral = "param assertion failed"]():
         cond = cond.__mlir_i1__(), message = msg.value
     ]()
     return
+
+
+@always_inline("nodebug")
+fn unreachable[msg: StringLiteral = "unreachable code"]():
```
Comment (Mogball) [inline stdlib/src/builtin/constrained.mojo:39]: This is an extremely misleading name for the function. It doesn't actually indicate unreachable code in the control-flow sense, which will lead to lots of misuse of the function. For example,


```mojo
fn make_true() -> Boo: return True

fn infinite_loop():
    while make_true():
        pass
    unreachable()
```

Will fail to compile

Comment (Benny-Nottonson) [reply stdlib/src/builtin/constrained.mojo:39]: I think that's more of a compiler edge case than a misnamed function.

Comment (Mogball) [reply stdlib/src/builtin/constrained.mojo:39]: The compiler can't interprocedurally reason about the function call and prune dead code in order to make the program "correct". That would make the optimization level affect whether code compiles

Comment (Mogball) [reply stdlib/src/builtin/constrained.mojo:39]: Plus, this function doesn't actually model a divergent function call. Type checking will still complain that a function like

```
fn foo() -> Int:
    unreachable()
```

Doesn't return an `Int` as expected.

Comment (Benny-Nottonson) [reply stdlib/src/builtin/constrained.mojo:39]: But when is an example when the above code would be functioning as intended? Why is there an unreachable in the first place?

Comment (Benny-Nottonson) [reply stdlib/src/builtin/constrained.mojo:39]: > Plus, this function doesn't actually model a divergent function call. Type checking will still complain that a function like
> 
> 
> 
> ```
> 
> fn foo() -> Int:
> 
>     unreachable()
> 
> ```
> 
> 
> 
> Doesn't return an `Int` as expected.

I'd argue that is a valid error for that function

Comment (Mogball) [reply stdlib/src/builtin/constrained.mojo:39]: The point is that the name "unreachable" suggests divergent control flow, which is not what the function implements. The function you implemented is a `static_assert(false)`, which is something else.

Comment (Benny-Nottonson) [reply stdlib/src/builtin/constrained.mojo:39]: Do you have an alternative name that is more correct, or different implementation?

**Comment by Brian-M-J** [comment]
> ```mojo
> contrained[value == 1 or value == 2]()
> ```

I assume you mean `constrained`?

**Comment by Benny-Nottonson** [comment]
> > ```mojo
> > contrained[value == 1 or value == 2]()
> > ```
> 
> I assume you mean `constrained`?

my mistake, fixed

**Pull request closed without merging**